src/main/java/com/squareup/javapoet/ParameterizedTypeName.java

```diff
@@ -87,7 +156,24 @@ public static ParameterizedTypeName get(ParameterizedType type) {
 
   /** Returns a parameterized type equivalent to {@code type}. */
   static ParameterizedTypeName get(ParameterizedType type, Map<Type, TypeVariableName> map) {
-    return new ParameterizedTypeName(ClassName.get((Class<?>) type.getRawType()),
-        TypeName.list(type.getActualTypeArguments(), map));
+    ParameterizedType toTest = type;
+    Deque<ParameterizedType> enclosingStack = new ArrayDeque<ParameterizedType>();
+    do {
+      enclosingStack.push(toTest);
+      toTest = (toTest.getOwnerType() instanceof ParameterizedType)
+          && !Modifier.isFinal(((Class<?>) type.getRawType()).getModifiers())
+          ? (ParameterizedType) toTest.getOwnerType() : null;
+    } while (toTest != null);
+
+    ParameterizedTypeName paramTypeName = null;
```

Comment:
nit: please don’t invent abbreviations like `param` for things. Either drop the term completely `typeName` or spell it out completely.